### PR TITLE
Port momentum distribution to batched drivers

### PR DESCRIPTION
**Pull request by jtkrogel**
## Proposed changes

This PR contains a port of the momentum distribution estimator to the new batched drivers.  

Tasks remaining:

- [x] Enable random number generator access for batched estimators
- [x] Enable trial wavefunction access for batched estimators
- [x] Statistical tests (manually) validating batched n(k) vs legacy implementation

## What type(s) of changes does this code introduce?

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop.

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

**Comment by PDoakORNL** [comment]
unit tests?  There isn't even a test file added with an instantiation test.

**Comment by jtkrogel** [comment]
I'm verifying statistical validity against the legacy implementation before any of that.

**Comment by jtkrogel** [comment]
Statistical tests all pass (serial, mpi/threaded):
```
n(k) orig
[1.91922013 0.44873453 0.44873453 0.44844139 0.44844139 0.44814747
 0.44814747 0.44807606 0.44807606 0.33214109 0.33214109 0.33108728
 0.33108728 0.33067993 0.33067993 0.01429943 0.01429943 0.01357385
 0.01357385 0.01345037 0.01345037 0.01337646 0.01337646 0.0101715
 0.0101715  0.00978179 0.00978179 0.00899401 0.00899401 0.00896847
 0.00896847 0.00888209 0.00888209 0.00882053 0.00882053 0.00867145
 0.00867145 0.00863097 0.00863097 0.00852863 0.00852863 0.00851691
 0.00851691 0.00848388 0.00848388 0.00840306 0.00840306]

n(k) batched
[1.91894789 0.44885341 0.44885341 0.448575   0.448575   0.44914446
 0.44914446 0.44711183 0.44711183 0.33175263 0.33175263 0.32982983
 0.32982983 0.33079952 0.33079952 0.01486382 0.01486382 0.0133028
 0.0133028  0.01423535 0.01423535 0.01407374 0.01407374 0.01012347
 0.01012347 0.00899164 0.00899164 0.00907284 0.00907284 0.00851357
 0.00851357 0.0091587  0.0091587  0.00955982 0.00955982 0.00868792
 0.00868792 0.00975386 0.00975386 0.00915181 0.00915181 0.00779539
 0.00779539 0.0099763  0.0099763  0.00831896 0.00831896]

significance of difference
[-0.33739404  0.09947974  0.09947974  0.10617789  0.10617789  0.79053891
  0.79053891 -0.76450476 -0.76450476 -0.3198806  -0.3198806  -0.92396054
 -0.92396054  0.1071796   0.1071796   0.60894     0.60894    -0.26927455
 -0.26927455  0.77990047  0.77990047  0.67971182  0.67971182 -0.048731
 -0.048731   -0.77836805 -0.77836805  0.08017774  0.08017774 -0.44646822
 -0.44646822  0.27062623  0.27062623  0.73759228  0.73759228  0.01559072
  0.01559072  1.09996053  1.09996053  0.63463733  0.63463733 -0.68006466
 -0.68006466  1.49294927  1.49294927 -0.07997782 -0.07997782]
```

**Comment by jtkrogel** [comment]
Ready for review.

**Comment by ye-luo** [comment]
Can we have a unit test?

**Comment by jtkrogel** [comment]
Happy to do that for constructor/clone.  If accumulate is needed, then we should talk more.  

The number of other classes that need to be built is large and this should be wrapped as a separate .cpp to reuse and simplify unit testing for estimators.

**Comment by ye-luo** [comment]
accumulate is the one need to be covered by test.

**Comment by PDoakORNL** [comment]
For the driver tests this is how I built up the basics I needed.  Not idea but not that bad.

#include "Particle/tests/MinimalParticlePool.h"
#include "QMCWaveFunctions/tests/MinimalWaveFunctionPool.h"
#include "QMCHamiltonians/tests/MinimalHamiltonianPool.h"

**Comment by jtkrogel** [comment]
At this stage, a deterministic integration test is less effort and more comprehensive.  I will go this route so I have some other time to give to the MS ahead of the Sept. 30 deadline.

File: src/Estimators/MomentumDistribution.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+class MomentumDistribution : public OperatorEstBase
+{
+public:
+  using LatticeType = PtclOnLatticeTraits::ParticleLayout_t;
+  using RealType    = QMCTraits::RealType;
+  using ComplexType = QMCTraits::ComplexType;
+  using ValueType   = QMCTraits::ValueType;
+  using PosType     = QMCTraits::PosType;
+
+  //data members set only during construction
```
Comment (ye-luo) [inline src/Estimators/MomentumDistribution.h:37]: Then all these members needs to be marked const. Probably myRNG is an exception.

**Review by ye-luo — COMMENTED** [review]
Is there a way to disable codecov comments temporarily? It makes review harder.

**Comment by ye-luo** [comment]
> At this stage, a deterministic integration test is less effort and more comprehensive. I will go this route so I have some other time to give to the MS ahead of the Sept. 30 deadline.

"a deterministic integration test is less effort." It is actually more error-prone and need fights with numeric sometimes.
"more comprehensive." deterministic integration test is still a coarse level test. test_SpinDensityNew.cpp like test can be more comprehensive.

**Comment by jtkrogel** [comment]
@PDoakORNL do you want deterministic test + more help w/ milestone or a unit test?  I don't have time for both.  You choose.

**Comment by PDoakORNL** [comment]
Is this a short term time crunch i.e. can the unit test be added after the MS?

**Comment by jtkrogel** [comment]
Yes, the time crunch is short term.  If the statistical validation I've performed already is enough for now, then I can add a unit test after the MS is submitted.

File: src/Estimators/MomentumDistribution.h (hunk 1 above)
Comment (jtkrogel) [reply src/Estimators/MomentumDistribution.h:37]: I removed myRNG (unused).  I've marked const where possible.  Some are set with computation within the constructor.

**Comment by jtkrogel** [comment]
@ye-luo I've addressed your comment.  Ready for merge.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
I add #3475 #3476 #3477 as follow-ups

**Pull request merged**